### PR TITLE
Forward host header for default origin.

### PR DESCRIPTION
**Pull request by jmcarp**
* Expose default origin
* Forward host for default origin
* Update docs

### CODE REVIEW

**Comment by jmcarp** [comment]
To explain a little more: this patch adds a default origin to the broker configuration, which would point to the cloud.gov apps load balancer. Users who want to add a domain to cloud.gov wouldn't pass the `origin` option--they'd add the domain via the cf cli. Users who want to point a domain elsewhere, like to an S3 bucket, will still set the `origin` option. The magical thing that I don't like is that we forward the host header if the origin is unset or set to the default origin, but not if the user specifies a non-default origin.

We could do better by allowing users to pass a list of headers to forward. If we do that, we can either default the list to ["Host"] (and allow users to default to an empty list for external origins) or default the list to [] (and allow users to pass ["Host"] or ["*"] for the default origin). I was going to do this originally, but I was thinking it might be confusing to users, and to use when we're doing support. WDYT @cnelson @jcscottiii ?

File: broker/broker.go [hunk 1]
```diff
@@ -219,15 +188,74 @@ func (b *CdnServiceBroker) Update(
...
+		return
+	}
+	if options.Domain == "" && options.Origin == "" {
+		err = errors.New("must be invoked with `domain` or `origin` keys")
+		return
+	}
+	return
+}
+
+func (b *CdnServiceBroker) getHeaders(options Options) []string {
```
Comment (jcscottiii) [inline broker/broker.go]: comment for this function please :)

**Comment by cnelson** [comment]
>  The magical thing that I don't like is that we forward the host header if the origin is unset or set to the default origin, but not if the user specifies a non-default origin. 

I think this is fine -- having the broker configure the distribution to work with cloud.gov when using that as an origin doesn't seem magic to me, it seems like the right thing to do.

>  I was going to do this originally, but I was thinking it might be confusing to users

I agree, I'd vote for not exposing that level of complexity until we have users demanding it.

File: broker/broker.go [hunk 2]
```diff
@@ -246,16 +246,12 @@ func (b *CdnServiceBroker) parseUpdateDetails(details map[string]interface{}) (o
 	if err != nil {
 		return
 	}
-	if options.Domain == "" && options.Origin == "" {
```
Comment (jcscottiii) [inline broker/broker.go]: don't you still need `domain`?

**Comment by jmcarp** [comment]
Tests passing on staging, so I'm dropping the [WIP]. If @jcscottiii and @cnelson, lmk if you have more comments, and I'll address them all together.

**Comment by jmcarp** [comment]
Also, thoughts on whether we should automatically create and destroy private domains on provision / unprovision?

**Comment by cnelson** [comment]
> Also, thoughts on whether we should automatically create and destroy private domains on provision / unprovision?

I vote no; less magic is better.  But WDYT about failing with a messaging tellig the user they need to create it if they specify a `domain` that CF doesn't know about?

**Comment by jmcarp** [comment]
I like that--will make that change once https://github.com/cloudfoundry-community/go-cfclient/pull/55 lands.

**Comment by jmcarp** [comment]
Updated! @cnelson, can you review the latest changes when you get a chance?

File: broker/broker.go [hunk 3]
```diff
@@ -219,15 +192,83 @@ func (b *CdnServiceBroker) Update(
...
+	if err != nil {
+		return
+	}
+	if options.Domain == "" {
+		err = errors.New("must pass non-empty `domain`")
+		return
+	}
+	if options.Origin == b.settings.DefaultOrigin {
+		if _, err = b.cfclient.GetDomainByName(options.Domain); err != nil {
+			err = fmt.Errorf("Domain %s does not exist; create it with `cf create-domain <my-organization> %s`", options.Domain, options.Domain)
```
Comment (cnelson) [inline broker/broker.go]: WDYT about putting the actual org namein the error?  It looks like the org guid is already in details so easy to lookup the name?

Comment (jmcarp) [inline broker/broker.go]: Looks like we're sending another patch to go-cfclient!

Comment (cnelson) [inline broker/broker.go]: lol how does it not have get-org-by-id :)

**Comment by jmcarp** [comment]
@cnelson: I fetched the org name when domains don't exist like you suggested, and acceptance tests haven't broken. I think it's ready!

**Comment by cnelson** [comment]
As soon as https://github.com/18F/cg-site/pull/752 gets merged :shipit: !

**Comment by jcscottiii** [comment]
@jmcarp one more question, the rails apps were having trouble with the `X-Forwarded-` header. will this allow users to modify that? (i see getHeaders returns either `Host` or empty slice.

**Comment by jmcarp** [comment]
@jcscottiii: the issue was that we weren't forwarding the original header, so that the header became the cloud.gov domain instead of the cloudfront domain. We could either forward the origin host header, or set x-forwarded-host. We wound up forwarding based on reasons in https://github.com/18F/cf-cdn-service-broker/issues/64.

**Pull request merged**